azavea/pfb-network-connectivity
# Support states/provinces for non-US cities

## Overview

Converts the `Neighborhood.state_abbrev` field from supporting only US states to supporting states/provinces for any country.  Whether we actually collect state/province for a given country depends on the `COUNTRY_CONFIG` in settings.  The front-end relies on the subdivision information included in the `countries` API endpoint, and the back-end validation checks that the selected country should have subdivisions and that the selected subdivision abbreviation is one of the valid ones.

### Demo
![image](https://user-images.githubusercontent.com/6598836/55816702-f01a8000-5ac0-11e9-93c8-1e1387c2692c.png)

### Notes

This doesn't change the labeling (see #733 for that), so the change isn't really perceptible on the public site.  On the admin site, though, the form has changed a bit--Country is before State, now, since having a lower question affect the display of an earlier one seems a bit confusing, and the State field will now stick around when Canada is selected. (see #735 for adding Province to the nomenclature)

## Testing Instructions

- Go to the [Create Neighborhood form](http://localhost:9301/#/admin/neighborhoods/create/).
   - Entering a US city should work as before (except with the fields in a different order). State should be required.
   - Entering a Canadian city should now require that you select the province.
   - Selecting any country besides those two should make the State field disappear.
   - If you pick a state then change to a different country that doesn't support states, the state should not get submitted with the save request.
- Go to the [API interface](http://localhost:9202/api/neighborhoods/) and try to submit an invalid neighborhood, such as
   - US or Canada with no state
   - US or Canada with an invalid state
   - Any other country with a value for state
- The public site Places list and details should be unchanged.

## Checklist


Resolves #732
